qdrant-landing/content/documentation/tutorials/retrieval-quality.md

```diff
@@ -117,111 +91,126 @@ client.upload_points(  # upload_points is available as of qdrant-client v1.7.1
 
-As we can see, the precision of the approximate search vs exact search is pretty high. There are, however, some scenarios when we
-need higher precision and can accept higher latency. HNSW is pretty tunable, and we can increase the precision by changing its parameters.
-  
-## Tweaking the HNSW parameters
+### Step 6: Tune HNSW Parameters for Better Precision
 
-HNSW is a hierarchical graph, where each node has a set of links to other nodes. The number of edges per node is called the `m` parameter. 
-The larger the value of it, the higher the precision of the search, but more space required. The `ef_construct` parameter is the number of 
-neighbours to consider during the index building. Again, the larger the value, the higher the precision, but the longer the indexing time.
-The default values of these parameters are `m=16` and `ef_construct=100`. Let's try to increase them to `m=32` and `ef_construct=200` and
-see how it affects the precision. Of course, we need to wait until the indexing is finished before we can perform the search.
+The HNSW algorithm used in Qdrant can be fine-tuned for better precision by adjusting parameters like `m` and `ef_construct`.
 
 ```python
+# Tune HNSW parameters for improved precision
 client.update_collection(
     collection_name="arxiv-titles-instructorxl-embeddings",
     hnsw_config=models.HnswConfigDiff(
-        m=32,  # Increase the number of edges per node from the default 16 to 32
-        ef_construct=200,  # Increase the number of neighbours from the default 100 to 200
+        m=32,          # Increase from the default 16 to 32
+        ef_construct=200,  # Increase from the default 100 to 200
     )
 )
 
+# Wait for re-indexing to complete
 while True:
     collection_info = client.get_collection(collection_name="arxiv-titles-instructorxl-embeddings")
     if collection_info.status == models.CollectionStatus.GREEN:
-        # Collection status is green, which means the indexing is finished
         break
+
+# Measure new precision
+print(f"New avg(precision@5) = {avg_precision_at_k(k=5)}")
 ```
 
-The same function can be used to calculate the average `precision@5`:
+By increasing `m` and `ef_construct`, you're allowing for a more connected graph and a more exhaustive search during indexing, which should lead to higher precision. After tuning, it's important to measure the precision again to verify improvements.
 
-```python
-print(f"avg(precision@5) = {avg_precision_at_k(k=5)}")
-```
+### How to Select HNSW Parameters
 
-Response:
+- If you require higher precision, increase `m` and `ef_construct` while considering the increased memory usage and indexing time.
+- If memory and indexing time are critical constraints, tune the parameters incrementally to find the right balance.
```

Comment:
By the way, these is also a third parameter : `ef` (also known as `efSearch`, it controls the number of neighbors evaluated during the search, a higher value may increase precision, however, it also increases latency